sparse/src/KokkosSparse_Utils.hpp

```diff
@@ -2330,6 +2330,163 @@ void validateCrsMatrix(int m, int n, const Rowmap &rowmapIn,
+ */
+template <typename crsMat_t>
+void kk_extract_diagonal_blocks_crsmatrix_sequential(
+    const crsMat_t &A, std::vector<crsMat_t> &DiagBlk_v) {
+  using row_map_type     = typename crsMat_t::row_map_type;
+  using entries_type     = typename crsMat_t::index_type;
+  using values_type      = typename crsMat_t::values_type;
+  using graph_t          = typename crsMat_t::StaticCrsGraphType;
+  using out_row_map_type = typename graph_t::row_map_type::non_const_type;
+  using out_entries_type = typename graph_t::entries_type::non_const_type;
+  using out_values_type  = typename crsMat_t::values_type::non_const_type;
+  using out_row_map_hostmirror_type = typename out_row_map_type::HostMirror;
+  using out_entries_hostmirror_type = typename out_entries_type::HostMirror;
+  using out_values_hostmirror_type  = typename out_values_type::HostMirror;
+
+  using ordinal_type = typename crsMat_t::non_const_ordinal_type;
+  using size_type    = typename crsMat_t::non_const_size_type;
+  using offset_view1d_type =
+      Kokkos::View<size_type *, Kokkos::LayoutLeft, Kokkos::HostSpace>;
+
+  row_map_type A_row_map = A.graph.row_map;
+  entries_type A_entries = A.graph.entries;
+  values_type A_values   = A.values;
+
+  auto A_row_map_h =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_row_map);
+  auto A_entries_h =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_entries);
+  auto A_values_h =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_values);
+
+  ordinal_type A_nrows  = static_cast<ordinal_type>(A.numRows());
+  ordinal_type A_ncols  = static_cast<ordinal_type>(A.numCols());
+  ordinal_type n_blocks = static_cast<ordinal_type>(DiagBlk_v.size());
+
+  if (A_nrows != A_ncols) {
+    std::ostringstream os;
+    os << "The diagonal block extraction only works with square matrices -- "
+          "matrix A: "
+       << A_nrows << " x " << A_ncols;
+    throw std::runtime_error(os.str());
+  }
+
+  if (n_blocks == 1) {
+    // One block case: simply shallow copy A to DiagBlk_v[0]
+    DiagBlk_v[0] = crsMat_t(A);
+  } else {
+    // n_blocks > 1
+    if (A_nrows == 0) {
+      // Degenerate case: A is an empty matrix
+      for (ordinal_type i = 0; i < n_blocks; i++) {
+        DiagBlk_v[i] = crsMat_t();
+      }
+    } else {
+      // A_nrows >= 1
+      ordinal_type rows_per_block = ((A_nrows % n_blocks) == 0)
+                                        ? (A_nrows / n_blocks)
+                                        : (A_nrows / n_blocks + 1);
+
+      std::vector<out_row_map_type> row_map_v(n_blocks);
+      std::vector<out_entries_type> entries_v(n_blocks);
+      std::vector<out_values_type> values_v(n_blocks);
+      std::vector<out_row_map_hostmirror_type> row_map_h_v(n_blocks);
+      std::vector<out_entries_hostmirror_type> entries_h_v(n_blocks);
+      std::vector<out_values_hostmirror_type> values_h_v(n_blocks);
+
+      ordinal_type row_start = 0;  // first row index of i-th diagonal block
```

Comment:
```suggestion
      ordinal_type blk_row_start = 0;  // first row index of i-th diagonal block
```

Reply:
@e10harvey Thanks for your suggestion. It was changed.